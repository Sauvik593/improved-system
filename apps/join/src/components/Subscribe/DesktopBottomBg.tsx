export const DesktopBottomBg = () => (
  <svg viewBox="0 0 1440 40" fill="none" xmlns="http://www.w3.org/2000/svg">
    <path d="M1440 0C714.542 69.4187 58.5527 30.3018 0 16.5283V0H1440Z" fill="#1f3b71" />
  </svg>
);
