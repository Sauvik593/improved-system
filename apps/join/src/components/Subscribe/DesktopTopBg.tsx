export const DesktopTopBg = () => (
  <svg viewBox="0 0 1440 48" fill="none" xmlns="http://www.w3.org/2000/svg">
    <path d="M1440 0C714.542 82.8493 58.5527 36.1644 0 19.726V48H1440V0Z" fill="#1f3b71" />
  </svg>
);
