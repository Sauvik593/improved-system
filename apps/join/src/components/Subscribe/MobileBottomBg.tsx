export const MobileBottomBg = () => (
  <svg viewBox="0 0 375 32" fill="none" xmlns="http://www.w3.org/2000/svg">
    <path d="M375 0C208.5 46.3198 20.5 29.736 0 26.3217V0H375Z" fill="#1f3b71" />
  </svg>
);
