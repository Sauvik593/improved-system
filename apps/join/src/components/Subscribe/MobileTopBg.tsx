export const MobileTopBg = () => (
  <svg viewBox="0 0 375 36" fill="none" xmlns="http://www.w3.org/2000/svg">
    <path d="M375 0C208.5 47.9278 20.5 30.7683 0 27.2355V36H375V0Z" fill="#1f3b71" />
  </svg>
);
