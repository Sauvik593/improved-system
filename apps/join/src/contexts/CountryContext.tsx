import { createContext } from 'react';
import type { Country } from '@lib/types';

export interface CountryContextParams {
  countries: Country[];
  country: Country | null;
}

export const CountryContext = createContext<CountryContextParams>({
  countries: [],
  country: null,
});
