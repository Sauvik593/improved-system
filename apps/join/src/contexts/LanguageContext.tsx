import { createContext } from 'react';

export interface LanguageContextParams {
  locale: string;
  locales: string[];
}

export const LanguageContext = createContext<LanguageContextParams>({
  locale: '',
  locales: [],
});
