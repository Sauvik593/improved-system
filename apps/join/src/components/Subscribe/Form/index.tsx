import { useContext } from 'react';
import type { UseFormRegister, FieldErrors } from 'react-hook-form';

import { useTranslation } from 'next-i18next';
import { startCase } from 'lodash';

import { LoadingButton } from '@components/common/LoadingButton';
import { BaseInput } from '@components/common/FormInputs/BaseInput';
import { HiddenInput } from '@components/common/FormInputs/HiddenInput';
import { LanguageContext } from '@contexts/LanguageContext';
import { CountryContext } from '@contexts/CountryContext';
import { NewsletterFormProps } from './submitForm';
import { NewsletterFooter } from '../NewsletterFooter';

interface Props {
  handleSubmit: () => void;
  register: UseFormRegister<NewsletterFormProps>;
  errors: FieldErrors<NewsletterFormProps>;
  submitting: boolean;
}

export const Form = ({ handleSubmit, register, errors, submitting }: Props) => {
  const { t } = useTranslation('common');
  const { locale } = useContext(LanguageContext);
  const { country } = useContext(CountryContext);
  return (
    <div data-testid="form-wrapper" className="lg:max-w-3xl">
      <h3 className="text-h-4-sm lg:text-h-4 mx-auto mb-4 text-center font-bold">
        {t('newsletter.title')}
      </h3>
      <form onSubmit={handleSubmit} data-testid="form">
        <div className="mx-auto mb-5 w-3/4 lg:grid lg:grid-cols-3 lg:gap-2">
          <div className="w-full lg:col-span-2 lg:mr-4">
            <BaseInput
              name="email"
              type="email"
              placeholder={t('newsletter.email.placeholder')}
              register={register}
              error={errors.email}
              testId="email"
              showTooltip
            />
          </div>
          <HiddenInput name="hsLanguage" register={register} value={locale} />
          <HiddenInput
            name="defaultNationContact"
            register={register}
            value={startCase(country?.translation_key as string)}
          />
          <HiddenInput name="receiveMarketAnalysis" register={register} value={'true'} />
          <div className="text-h-5-sm lg:text-h-5 mt-4 lg:mt-0 lg:grid-cols-1">
            <LoadingButton
              color="sky"
              size="normal"
              loading={submitting}
              label={t('ui.buttons.subscribe')}
            />
          </div>
        </div>
        <NewsletterFooter />
      </form>
    </div>
  );
};
