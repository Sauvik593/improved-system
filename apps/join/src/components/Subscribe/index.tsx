import { useForm } from 'react-hook-form';

import { useTranslation } from 'next-i18next';

import cn from 'classnames';

import { yupResolver } from '@hookform/resolvers/yup';

import { useHubspotForm } from '@hooks/useHubspotForm';

import { DesktopTopBg } from './DesktopTopBg';
import { MobileTopBg } from './MobileTopBg';
import { DesktopBottomBg } from './DesktopBottomBg';
import { MobileBottomBg } from './MobileBottomBg';
import { createEmailValidationSchema } from './Form/createEmailValidationSchema';
import { submitForm, NewsletterFormProps } from './Form/submitForm';
import { Form } from './Form';
import { SuccessView } from './Form/SuccessView';
import { ErrorView } from './Form/ErrorView';

interface Props {
  className?: string;
}

export const Subscribe = ({ className }: Props) => {
  const { t } = useTranslation('common');
  const {
    register,
    handleSubmit,
    getValues,
    formState: { errors, isSubmitting },
  } = useForm<NewsletterFormProps>({
    resolver: yupResolver(createEmailValidationSchema(t)),
  });

  const { error, setFormState, submit, success } = useHubspotForm({
    formService: () => submitForm(getValues()),
  });

  const renderView = () => {
    switch (true) {
      case success:
        return <SuccessView email={getValues('email')} />;
      case error:
        return <ErrorView onRetry={() => setFormState('form')} />;
      default:
        return (
          <Form
            register={register}
            errors={errors}
            handleSubmit={handleSubmit(submit)}
            submitting={isSubmitting}
          />
        );
    }
  };

  return (
    <section className={cn('relative w-full', className)}>
      <div className="relative hidden w-full lg:block xl:-mb-[1px]">
        <DesktopTopBg />
      </div>
      <div className="relative block w-full lg:hidden">
        <MobileTopBg />
      </div>
      <div className="bg-[#1f3b71] px-4 pt-10 pb-2 md:px-0">
        <div className="flex justify-center text-white">{renderView()}</div>
      </div>
      <div className="relative hidden w-full bg-white pb-4 lg:block">
        <DesktopBottomBg />
      </div>
      <div className="relative block w-full bg-white pb-6 lg:hidden">
        <MobileBottomBg />
      </div>
    </section>
  );
};
