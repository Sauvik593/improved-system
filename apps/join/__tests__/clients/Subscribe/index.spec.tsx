import {
  render,
  screen,
  waitFor,
  getByTestId,
  fireEvent,
  act,
  RenderResult,
} from '@testing-library/react';
import { rest } from 'msw';
import { setupServer } from 'msw/node';

import { setupIntersectionObserverMock } from '../../intersectionObserverMock';
import { CountryContext } from '@contexts/CountryContext';
import { LanguageContext } from '@contexts/LanguageContext';
import { Subscribe } from '@components/Subscribe';
import { countrySpain, mockedCountries } from '../../countriesMock';

jest.mock('react-i18next', () => ({
  useTranslation: () => ({ t: (key: string) => key }),
}));

const server = setupServer(
  rest.post('/new-join/api/hubspot', (_req, res, ctx) => res(ctx.status(200))),
);

const submitForm = async (values: { email: string }): Promise<void> => {
  const formWrapper = screen.getByTestId('form-wrapper');
  const emailField = getByTestId(formWrapper, 'email') as HTMLInputElement;
  const form = getByTestId(formWrapper, 'form') as HTMLFormElement;

  await act(async () => {
    await waitFor(async () => {
      fireEvent.change(emailField, { target: { value: values.email } });
      fireEvent.submit(form);
    });
  });
};

describe('Subscribe', () => {
  let renderResult: RenderResult;

  beforeEach(() => {
    renderResult = render(
      <CountryContext.Provider value={{ country: countrySpain, countries: mockedCountries }}>
        <LanguageContext.Provider value={{ locale: 'en', locales: ['en', 'es', 'pt', 'fr', 'it'] }}>
          <Subscribe />
        </LanguageContext.Provider>
      </CountryContext.Provider>,
    );
    setupIntersectionObserverMock();
  });

  beforeAll(() => {
    server.listen();
  });

  afterEach(() => {
    server.resetHandlers();
  });

  afterAll(() => server.close());

  describe('When initially rendered', () => {
    it('displays the form view', () => {
      const successView = screen.queryByTestId('success-view');
      const errorView = screen.queryByTestId('error-view');

      expect(successView).not.toBeInTheDocument();
      expect(errorView).not.toBeInTheDocument();
    });
  });

  describe('When the form is submitted', () => {
    it('displays the form success', async () => {
      await submitForm({
        email: 'test@test.com',
      });

      await waitFor(() => {
        expect(screen.queryByTestId('success-view')).toBeInTheDocument();
        expect(screen.queryByTestId('error-view')).not.toBeInTheDocument();
      });
    });
  });

  describe('When the form is submitted but request fails', () => {
    it('displays the error modal', async () => {
      server.use(rest.post('/new-join/api/hubspot', (req, res, ctx) => res(ctx.status(500))));

      await submitForm({
        email: 'test@test.com',
      });

      await waitFor(() => {
        expect(screen.queryByTestId('error-view')).toBeInTheDocument();
        expect(screen.queryByTestId('success-view')).not.toBeInTheDocument();
      });
    });
  });
  describe('When form is filled', () => {
    it('shows validation errors', async () => {
      const { asFragment } = renderResult;

      await submitForm({
        email: 'not_valid_email',
      });

      await waitFor(() => {
        expect(screen.queryByTestId('email-error')).toBeInTheDocument();
        expect(asFragment()).toMatchSnapshot();
      });
    });
  });
});
