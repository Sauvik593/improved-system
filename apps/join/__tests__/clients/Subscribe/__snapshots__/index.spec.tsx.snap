// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Subscribe When form is filled shows validation errors 1`] = `
<DocumentFragment>
  <section
    class="relative w-full"
  >
    <div
      class="relative hidden w-full lg:block xl:-mb-[1px]"
    >
      <svg
        fill="none"
        viewBox="0 0 1440 48"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M1440 0C714.542 82.8493 58.5527 36.1644 0 19.726V48H1440V0Z"
          fill="#1f3b71"
        />
      </svg>
    </div>
    <div
      class="relative block w-full lg:hidden"
    >
      <svg
        fill="none"
        viewBox="0 0 375 36"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M375 0C208.5 47.9278 20.5 30.7683 0 27.2355V36H375V0Z"
          fill="#1f3b71"
        />
      </svg>
    </div>
    <div
      class="bg-[#1f3b71] px-4 pt-10 pb-2 md:px-0"
    >
      <div
        class="flex justify-center text-white"
      >
        <div
          class="lg:max-w-3xl"
          data-testid="form-wrapper"
        >
          <h3
            class="text-h-4-sm lg:text-h-4 mx-auto mb-4 text-center font-bold"
          >
            newsletter.title
          </h3>
          <form
            data-testid="form"
          >
            <div
              class="mx-auto mb-5 w-3/4 lg:grid lg:grid-cols-3 lg:gap-2"
            >
              <div
                class="w-full lg:col-span-2 lg:mr-4"
              >
                <div>
                  <input
                    class="border-sierra-night-40 text-sierra-night-60 w-full rounded-md border p-2 pl-4"
                    data-testid="email"
                    name="email"
                    placeholder="newsletter.email.placeholder"
                    type="email"
                  />
                  <p
                    class="px-2 error-tooltip"
                    data-testid="email-error"
                  >
                    newsletter.email.format
                  </p>
                </div>
              </div>
              <input
                name="hsLanguage"
                type="hidden"
                value="en"
              />
              <input
                name="defaultNationContact"
                type="hidden"
                value="Spain"
              />
              <input
                name="receiveMarketAnalysis"
                type="hidden"
                value="true"
              />
              <div
                class="text-h-5-sm lg:text-h-5 mt-4 lg:mt-0 lg:grid-cols-1"
              >
                <button
                  class="inline-block w-full rounded-full px-4 text-center font-bold text-white disabled:cursor-not-allowed bg-sky-100 border-sky-100 text-tile-100 hover:border-sky-150 hover:bg-sky-150 focus:border-sky-150 focus:bg-sky-150 border-1 py-2"
                >
                  ui.buttons.subscribe
                </button>
              </div>
            </div>
            <footer>
              <div
                class="text-center font-medium"
              >
                newsletter.footer.text
                <a
                  class="mx-1 underline"
                  href="https://docs.kyero.com/en/privacy"
                >
                  newsletter.footer.privacy_policy
                </a>
                <span>
                  &
                </span>
                <a
                  class="ml-1 underline"
                  href="https://docs.kyero.com/en/terms"
                >
                  newsletter.footer.terms
                </a>
              </div>
            </footer>
          </form>
        </div>
      </div>
    </div>
    <div
      class="relative hidden w-full bg-white pb-4 lg:block"
    >
      <svg
        fill="none"
        viewBox="0 0 1440 40"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M1440 0C714.542 69.4187 58.5527 30.3018 0 16.5283V0H1440Z"
          fill="#1f3b71"
        />
      </svg>
    </div>
    <div
      class="relative block w-full bg-white pb-6 lg:hidden"
    >
      <svg
        fill="none"
        viewBox="0 0 375 32"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M375 0C208.5 46.3198 20.5 29.736 0 26.3217V0H375Z"
          fill="#1f3b71"
        />
      </svg>
    </div>
  </section>
</DocumentFragment>
`;
